content/interoperability.tex

```diff
@@ -0,0 +1,191 @@
+call to \FUNC{MPI\_Finalize}.
+
+\apiimpnotes{
+Portable implementations of OpenSHMEM and MPI must ensure that the initialization
+calls can be made in an arbitrary order within a program; the same rule also
+applies to the finalization calls. A software runtime that utilizes a shared
+communication resource for \openshmem and MPI communication may maintain an
+internal reference counter in order to ensure that the shared resource is
+initialized only once and thus no shared resource is released until the last
+finalization call is made.
+}
+
+
+\subsection{Dynamic Process Creation}
+\label{subsec:interoperability:mpmd}
+
+MPI defines a dynamic process model that allows creation of processes after
+an MPI application has started (e.g., by calling \FUNC{MPI\_Comm\_spawn}) and
+connection to independent processes (e.g., through \FUNC{MPI\_Comm\_accept}
+and \FUNC{MPI\_Comm\_connect})
+and provides a mechanism to establish communication
+between the newly created processes and the existing MPI application (see
+MPI standard version 3.1, Chapter 10).
+Unlike MPI, \openshmem starts all processes at once and requires all PEs to
+collectively allocate and initialize resources (e.g., symmetric heap) used by
+the \openshmem library before any other \openshmem routine may
+be called. Communicating with a dynamically created process in the \openshmem
+environment may result in undefined behavior.
+Hence, users should not use \openshmem and MPI dynamic process models
+in the same program.
+
+
+\subsection{Thread Safety}
+\label{subsec:interoperability:thread}
+Both \openshmem and MPI define the interaction with user threads in a program
+with routines that can be used for initializing and querying the thread
+environment. In a hybrid program, the user may request different thread levels
+at the initialization calls of \openshmem and MPI environments; however, the
+returned support level provided by the \openshmem library might be different
+from that returned in an \openshmem-only program. For instance, the former
+initialization call in a hybrid program may initialize a resource with the
+user-requested thread level, but the supported level cannot be updated by the latter
+initialization call if the underlying software runtime of \openshmem and MPI
+share the same internal communication resource.
+The program should always check the \VAR{provided} thread level returned
+at the corresponding initialization call or query the level of thread support
+after initialization to portably ensure thread support in each communication
+environment.
+
+Both \openshmem and MPI define similar thread levels, namely, \VAR{THREAD\_SINGLE},
+\VAR{THREAD\_FUNNELED}, \VAR{THREAD\_SERIALIZED}, and \VAR{THREAD\_MULTIPLE}.
+When requesting threading support in a hybrid program, however,
+the following additional rules are applied if the implementations of \openshmem
+and MPI share the same internal communication resource.
```

Comment:
Is it possible to query whether synchronization requirements are combined for both MPI and SHMEM or separate (i.e. separate serialization domains)?

Reply:
Good idea. We may also add a query property to indicate whether the same master thread is required for the `THREAD_FUNNELED` safety. But I need time to think about a use case which can benefit from the feature. Maybe it is better to first wrap up the current proposal and propose the other properties separately. What do you think?

Reply:
Over the WG call, we decided to postpone the query API into a separate proposal and we will work on the thread safety properties there.